File: NeetCode 150/Backtracking/90. Subsets II.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        #subset 1 has unique elements. Here we have duplicate elements
        nums.sort() # Sort to group duplicates together
        res = []
        subset = []

        def backtrack(start):
            # Every path is a valid subset
            res.append(subset[:])

            for i in range(start,len(nums)):
                # Skip duplicates at the same recursion level
                if i > start and nums[i] == nums[i-1]:
                    continue
                subset.append(nums[i])
                backtrack(i+1)
                subset.pop()

        backtrack(0)
        return res

File: NeedCode 150 (JAVA)/Backtracking/90. Subsets II.java
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    private List<Integer> subset = new ArrayList<>();
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        backtrack(0,nums);
        return res;
    }

    private void backtrack(int start, int[] nums){
        //every path is valid path
        res.add(new ArrayList<>(subset));

        for(int i = start;i<nums.length;i++){
            //skip duplicates at same recursion level
            if(i>start && nums[i] == nums[i-1]) continue;

            subset.add(nums[i]);
            backtrack(i+1,nums);
            subset.remove(subset.size() -1 );

        }
    }
}
